radixdlt-core/radixdlt/src/main/java/com/radixdlt/api/service/SubmissionService.java

```diff
@@ -115,14 +118,38 @@ private TxnConstructionRequest toConstructionRequest(
 			var success = completableFuture.get();
 			return Result.ok(success.getTxn().getId());
 		} catch (ExecutionException e) {
-			logger.warn("Unable to fulfill submission request: " + txn.getId() + ": ", e);
-			return SUBMISSION_FAILURE.with(e.getMessage()).result();
+			var cause = lookupCause(e);
+
+			logger.warn("Unable to fulfill submission request for TxID (" + txn.getId() + ")", e);
+			return SUBMISSION_FAILURE.with(cause.getMessage()).result();
 		} catch (InterruptedException e) {
 			Thread.currentThread().interrupt();
 			throw new IllegalStateException(e);
 		}
 	}
 
+	private Throwable lookupCause(Throwable e) {
+		var reportedException = e;
+
+		while (reportedException.getCause() instanceof MempoolRejectedException) {
+			reportedException = reportedException.getCause();
+		}
+
+		while (reportedException.getCause() instanceof RadixEngineException) {
+			reportedException = reportedException.getCause();
+		}
+
+		while (reportedException.getCause() instanceof ConstraintMachineException) {
+			reportedException = reportedException.getCause();
+		}
+
+		if (reportedException instanceof  ConstraintMachineException && reportedException.getCause() != null) {
+			reportedException = reportedException.getCause();
+		}
+
+		return reportedException;
```

Comment:
Can we report the full exception stack to end user?

Reply:
Agree, this would probably be most helpful

Reply:
I disagree. User wants to know the real cause rather than dig through full stack trace (which is quite long in this case). The full stack trace is provided in the logs.

Reply:
I think it depends on what you want to achieve: 
1. To provide failure cause to the end users so that they know what to do next;
2. To provide a trace of failures, so that app developers can diagnose the problem without consulting the node.

For 1, we need to provide explicit list of error codes with documentation on the suggested solution.
For 2, it's better to provide as much details as possible IMHO.

What do you think?

Reply:
It's a user-facing API (for example, wallet uses it), so the aim to achieve #1. Unfortunately, at the point where I'm intercepting the exception, I have no detailed information such as error code. Actually, the whole method is an attempt to retrieve some human-recognizable information instead of just the transaction blob which we did return previously.
Also, it was unexpected, but I've discovered that messages returned by this method are much more helpful even for the developer: simple text search with the error cause immediately points to the correct location in code.